File: src/main/java/todayquest/quest/entity/Quest.java
package todayquest.quest.entity;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import todayquest.common.BaseTimeEntity;
import todayquest.quest.dto.QuestRequestDto;
import todayquest.reward.entity.Reward;
import todayquest.user.entity.UserInfo;

import javax.persistence.*;
import java.time.LocalDate;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;

import static javax.persistence.FetchType.LAZY;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
public class Quest extends BaseTimeEntity {

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "quest_id")
    private Long id;

    @Column(length = 50, nullable = false)
    private String title;

    @Column(length = 300)
    private String description;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "user_id")
    private UserInfo user;

    @Column(name = "user_quest_seq")
    private Long seq;

    /**
     * QuestType 정보와 조합해 반복 주기를 설정한다.
     * 마찬가지로 배치 처리가 필요할 듯 하다.
     */
    @Column(nullable = false)
    private boolean isRepeat;

    /**
     * 현재 일간 퀘스트만 사용하므로 시간 정보만 받아서 처리, 날짜 정보는 앞단에서 오늘 날짜와 동일하게 처리한다.
     * 추후 주간, 월간, 연간 퀘스트 추가 시 이 필드에 날짜 정보를 입력 받아 배치 처리에 사용한다.
     */
    private LocalDate deadLineDate;
    private LocalTime deadLineTime;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private QuestState state;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private QuestType type;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private QuestDifficulty difficulty;

    @OneToMany(mappedBy = "quest")
    private List<QuestReward> rewards = new ArrayList<>();


    @Builder
    public Quest(String title, String description, UserInfo user, Long seq, boolean isRepeat, LocalDate deadLineDate, LocalTime deadLineTime,
                 QuestState state, QuestType type, QuestDifficulty difficulty) {

        this.title = title;
        this.description = description;
        this.user = user;
        this.seq = seq;
        this.isRepeat = isRepeat;
        this.deadLineDate = deadLineDate;
        this.deadLineTime = deadLineTime;
        this.state = state;
        this.type = type;
        this.difficulty = difficulty;
    }

    public List<QuestReward> updateQuestEntity(QuestRequestDto dto, List<Reward> updateRewards) {
        this.title = dto.getTitle();
        this.description = dto.getDescription();
        this.isRepeat = dto.isRepeat();
        this.deadLineDate = dto.getDeadLineDate();
        this.deadLineTime = dto.getDeadLineTime();
        this.difficulty = dto.getDifficulty();
        return updateRewardList(updateRewards);
    }

    private List<QuestReward> updateRewardList(List<Reward> updateRewards) {
        List<QuestReward> newRewards = new ArrayList<>();

        int updateCount = updateRewards.size();

        for (int i = 0; i < updateCount; i++) {
            QuestReward newReward = QuestReward.builder().reward(updateRewards.get(i)).quest(this).build();
            try {
                rewards.get(i).updateReward(updateRewards.get(i));
            } catch (IndexOutOfBoundsException e) {
                newRewards.add(newReward);
            }
        }

        int overCount = rewards.size() - updateCount;
        if (overCount > 0) {
            for (int i = updateCount; i < updateCount + overCount; i++) {
                rewards.remove(i);
            }
        }

        return newRewards;
    }

    public void changeState(QuestState state) {
        this.state = state;
    }

}

File: src/main/java/todayquest/quest/entity/Quest.kt
package todayquest.quest.entity

import jakarta.persistence.*
import org.springframework.security.access.AccessDeniedException
import todayquest.common.BaseTimeEntity
import todayquest.common.MessageUtil
import todayquest.quest.dto.*
import todayquest.user.entity.UserInfo

@Entity
class Quest(
    title: String,
    description: String?,
    user: UserInfo,
    seq: Long,
    state: QuestState = QuestState.PROCEED,
    type: QuestType,
) : BaseTimeEntity() {

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "quest_id")
    val id: Long = 0

    @Column(length = 50, nullable = false)
    var title: String = title
        protected set

    @Column(length = 300)
    var description: String? = description
        protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    val user: UserInfo = user

    @Column(name = "user_quest_seq", nullable = false)
    val seq: Long = seq

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val type: QuestType = type

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var state: QuestState = state
        protected set

    @OneToMany(mappedBy = "quest", cascade = [CascadeType.ALL], orphanRemoval = true)
    private val _detailQuests: MutableList<DetailQuest> = mutableListOf()
    val detailQuests : List<DetailQuest>
        get() = _detailQuests.toList()

    fun updateQuestEntity(dto: QuestRequest) {
        title = dto.title
        description = dto.description
        updateDetailQuests(dto.details)
    }

    fun updateDetailQuests(detailRequests: List<DetailRequest>) {
        val newDetailQuests: MutableList<DetailQuest> = mutableListOf()

        val updateCount = detailRequests.size
        for (i in 0 until updateCount) {
            val newDetailQuest = detailRequests[i]
            try {
                _detailQuests[i].updateDetailQuest(newDetailQuest)
            } catch (e: IndexOutOfBoundsException) {
                newDetailQuests.add(newDetailQuest.mapToEntity(this))
            }
        }

        val overCount: Int = _detailQuests.size - updateCount
        if (overCount > 0) {
            for (i in updateCount until updateCount + overCount) {
                _detailQuests.removeAt(updateCount)
            }
        }

        _detailQuests.addAll(newDetailQuests)
    }

    fun completeQuest() {
        require(state != QuestState.DELETE) { MessageUtil.getMessage("quest.error.deleted") }
        require(state == QuestState.PROCEED) { MessageUtil.getMessage("quest.error.not-proceed") }
        require(canComplete()) { MessageUtil.getMessage("quest.error.complete.detail") }

        state = QuestState.COMPLETE
    }

    fun deleteQuest() {
        state = QuestState.DELETE
    }

    fun discardQuest() {
        require(state != QuestState.DELETE) { MessageUtil.getMessage("quest.error.deleted") }
        require(state == QuestState.PROCEED) { MessageUtil.getMessage("quest.error.not-proceed") }

        state = QuestState.DISCARD
    }

    fun failQuest() {
        state = QuestState.FAIL
    }

    fun checkIsProceedingQuest() {
        require(state == QuestState.PROCEED) { MessageUtil.getMessage("quest.error.not-proceed") }
    }

    fun checkIsQuestOfValidUser(userId: Long) {
        if (user.id != userId)
            throw AccessDeniedException(MessageUtil.getMessage("exception.access.denied"))
    }

    fun canComplete(): Boolean {
        return detailQuests.stream()
            .allMatch(DetailQuest::isCompletedDetailQuest)
    }

    fun isMainQuest(): Boolean {
        return type == QuestType.MAIN
    }

    fun interactWithDetailQuest(detailQuestId: Long, request: DetailInteractRequest? = null): DetailResponse {
        val detailQuest = _detailQuests.firstOrNull { it.id == detailQuestId }
            ?: throw IllegalArgumentException(MessageUtil.getMessage("exception.badRequest"))

        checkIsProceedingQuest()

        if(request != null) {
            detailQuest.changeCount(request.count)
            return DetailResponse.createDto(detailQuest, canComplete())
        }

        if(detailQuest.isCompletedDetailQuest()) {
            detailQuest.resetCount()
            return DetailResponse.createDto(detailQuest)
        }

        detailQuest.addCount()

        return DetailResponse.createDto(detailQuest, canComplete())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Quest

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

}
